examples/universe.m31

```diff
@@ -0,0 +1,106 @@
+(* An example showing how we might work with a universe,
+   where conversion from names to types is done automatically
+   through handlers. *)
+
+constant U : Type
+constant El : U → Type
+
+(* The name of a product *)
+constant pi : ∏ (a : U), (El a → U) → U
+constant El_pi : ∏ (a : U) (b : El a → U), El (pi a b) ≡ (∏ (x : El a), El (b x))
+now betas = add_beta El_pi
+
+(* The name of an arrow *)
+constant arrow : U → U → U
+constant arrow_def : ∏ (a b : U), arrow a b ≡ pi a (λ _, b)
+now betas = add_beta arrow_def
+
+(* The name of an equality type *)
+constant eq : ∏ (a : U), El a → El a → U
+constant El_eq : ∏ (a : U) (x y : El a), El (eq a x y) ≡ (x ≡ y)
+now betas = add_beta El_eq
+
+(* The unit type *)
+constant Unit : Type
+constant tt : Unit
+constant Unit_eta : ∏ (x y : Unit), x ≡ y
+now etas = add_eta Unit_eta
+
+(* The name of the unit type *)
+constant unit : U
+constant El_unit : El unit ≡ Unit
+now betas = add_beta El_unit
+
+let whnf_e e = match whnf e with |- _ : _ == ?e => e end
+
+(* Reverse-engineer the name of a type with a proof that
+   it is correctly reverse-engineered *)
+let rec name_of_eq T =
+  match T with
+  | ⊢ El _ => Some (refl T)
+  | ⊢ Unit => Some El_unit
+  | ⊢ (?x ≡ ?y) =>
+      let A = typeof x in
+      match name_of_eq A with
+      | None => None
+      | Some (⊢ ?ξ : El ?a ≡ _) => 
+        let x' = convert x (symmetry ξ)
+        and y' = convert y (symmetry ξ) in
+        let α = (El_eq a x' y') in
+        let β = convert_eq (x' ≡ y') ξ  in
+        Some (transitivity α β)
+      end
+  | ⊢ Π (?x : ?A), ?B => 
+      match name_of_eq A with
+      | None => None
+      | Some (⊢ ?ξ : El ?a ≡ _) => 
+          match name_of_eq B with
+          | None => None
+          | Some (⊢ ?ξ2 : El ?b ≡ _) => 
+             Some (El_pi a (λ (y : El a), b where x = y))
+          end
+      end
+  | ?J => 
+    let ζ = whnf T  in
```

Comment:
I think this could be:

     match whnf T with
     | ⊢ ?ζ : _ ≡ ?T' =>
       match T' with
       | _ => None
      | _ => ...
       end
     end